modules/dns-helper/dns_local_server.go

```diff
@@ -0,0 +1,160 @@
+// it uses a handler that will send replies stored in their internal DNSDatabase
+func setupTestDNSServers() (s1, s2 *dnsTestServer) {
+	s1 = runTestDNSServer("0")
+	s2 = runTestDNSServer("0")
+
+	q := DNSQuery{"NS", testDomain}
+	a := DNSAnswers{{"NS", s1.Address() + "."}, {"NS", s2.Address() + "."}}
+	s1.AddEntryToDNSDatabase(q, a)
+	s2.AddEntryToDNSDatabase(q, a)
+
+	s1.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		stdDNSHandler(w, r, s1, false)
+	})
+	s2.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		stdDNSHandler(w, r, s2, true)
+	})
+
+	return s1, s2
+}
+
+// setupTestDNSServersRetry runs and returns 2x local dnsTestServer, initialized with NS records for the testDomain pointing to themselves
+// it uses a handler that will send replies stored in their internal DNSDatabase, and then switch to their DNSDatabaseRetry after some time
+func setupTestDNSServersRetry() (s1, s2 *dnsTestServer) {
+	s1 = runTestDNSServer("0")
+	s2 = runTestDNSServer("0")
+
+	q := DNSQuery{"NS", testDomain}
+	a := DNSAnswers{{"NS", s1.Address() + "."}, {"NS", s2.Address() + "."}}
+	s1.AddEntryToDNSDatabase(q, a)
+	s2.AddEntryToDNSDatabase(q, a)
+	s1.AddEntryToDNSDatabaseRetry(q, a)
+	s2.AddEntryToDNSDatabaseRetry(q, a)
+
+	s1.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		retryDNSHandler(w, r, s1, false)
+	})
+	s2.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		retryDNSHandler(w, r, s2, true)
+	})
+
+	return s1, s2
+}
+
+// runTestDNSServer starts and returns a new dnsTestServer listening in localhost and the given UDP port
+func runTestDNSServer(port string) *dnsTestServer {
+	listener, err := net.ListenPacket("udp", "127.0.0.1:"+port)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	mux := dns.NewServeMux()
+	server := &dns.Server{PacketConn: listener, Net: "udp", Handler: mux}
+
+	go func() {
+		if err := server.ActivateAndServe(); err != nil {
+			log.Printf("Error in local DNS server: %s", err)
+		}
+	}()
+
+	return newDNSTestServer(server)
+}
+
+// doDNSAnswer sends replies to the DNS question from client, using the dnsDatabase to lookup the answers to the query
+// when invertAnswers is true, reverses the order of the answers from the dnsDatabase, useful to simulate realistic nameservers behaviours
+func doDNSAnswer(w dns.ResponseWriter, r *dns.Msg, d dnsDatabase, invertAnswers bool) {
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative = true
+	q := m.Question[0]
+	qtype := dns.TypeToString[q.Qtype]
+	answers := d[DNSQuery{qtype, strings.TrimSuffix(q.Name, ".")}]
+
+	var seen = make(map[DNSAnswer]bool)
+
+	for _, r := range answers {
+		if seen[r] {
+			continue
+		}
+		seen[r] = true
+
+		rr, err := dns.NewRR(fmt.Sprintf("%s %s", q.Name, r.String()))
+
+		if err != nil {
+			log.Fatalf("err: %s", err)
```

Comment:
Same as above about avoiding `log.Fatal`.

Reply:
Switch from `log.Fatalf` to `t.Fatalf`